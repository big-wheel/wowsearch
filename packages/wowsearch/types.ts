/**
 * @file types
 * @author Cuttle Cong
 * @date 2019/9/29
 *
 */

export * from 'wowsearch-parse/types/Config'
