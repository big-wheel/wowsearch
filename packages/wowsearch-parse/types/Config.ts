/**
 * @file Config
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */
import * as w from 'walli'
import * as each from 'lodash.foreach'
import * as clone from 'lodash.clonedeep'
import { Rule, walliRule } from '../match'
import { normalizeSelector } from '../selectVal'

type CommonConfig = {
  strip_chars?: string
  anchor_attribute_name?: string
  anchor_selector?: Selector
}

export type Selector = string | StrictSelector

export type StrictSelector = CommonConfig & {
  selector: string
  type?: 'xpath' | 'css'
  global?: boolean
  anchor_attribute_name?: string
  anchor_selector?: AnchorSelector | string
  default_value?: any
}

export type AnchorSelector = CommonConfig & {
  selector: string
  type?: 'xpath' | 'css'
  global?: boolean
  default_value?: any
}

const walliSelector = w.oneOf([
  w.string,
  w.leq({
    selector: w.string,
    type: w.oneOf(['xpath', 'css']).optional,
    default_value: w.any.optional,
    strip_chars: w.string.optional
  })
])

export type Selectors = {
  lvl0?: Selector
  lvl1?: Selector
  lvl2?: Selector
  lvl3?: Selector
  lvl4?: Selector
  lvl5?: Selector
  lvl6?: Selector
  text?: Selector
}

export type CrawlerConfig = CommonConfig & {
  js_render?: boolean
  js_waitfor?: string | number | Function
  start_urls?: Array<Rule | { url: Rule }>
  stop_urls?: Rule[]

  selectors: Selectors

  selectors_exclude?: string[]
  smart_crawling?: boolean
  force_crawling_urls?: boolean
}

export type Config = CrawlerConfig & {
  sitemap_urls?: string[]
  sitemap_urls_patterns?: Rule[]
  force_sitemap_urls_crawling?: boolean
}

const WalliDef = w.leq({
  js_render: w.boolean.optional,
  js_waitfor: w.oneOf([w.string, w.number, w.function_]).optional,
  strip_chars: w.string.optional,
  start_urls: w.arrayOf(w.oneOf([walliRule, w.leq({ url: w.string })]))
    .optional,
  stop_urls: w.arrayOf(walliRule).optional,

  selectors: {
    lvl0: walliSelector.optional,
    lvl1: walliSelector.optional,
    lvl2: walliSelector.optional,
    lvl3: walliSelector.optional,
    lvl4: walliSelector.optional,
    lvl5: walliSelector.optional,
    lvl6: walliSelector.optional,
    text: walliSelector.optional
  },

  selectors_exclude: w.arrayOf(w.string).optional
})

export function normalize(config: Config) {
  let errorMsg = WalliDef.toUnlawfulString(config)
  if (errorMsg) {
    throw new TypeError(errorMsg)
  }
  config = clone(config)

  config = Object.assign(
    {
      strip_chars: ' .,;:§¶',
      js_render: false,
      js_waitfor: 0,
      anchor_selector: 'a[id]',
      anchor_attribute_name: 'id',
      start_urls: [/.*/],
      stop_urls: [],
      selectors_exclude: [],
      sitemap_urls: [],
      sitemap_urls_patterns: [/.*/],
      force_sitemap_urls_crawling: false,
      smart_crawling: false,
      force_crawling_urls: false
    },
    config
  )

  each(config.selectors, (value, key) => {
    if (typeof value === 'string') {
      value = {
        selector: value
      }
    }
    if (value) {
      config.selectors[key] = Object.assign(
        {
          type: 'css',
          strip_chars: config.strip_chars,
          anchor_selector: config.anchor_selector,
          anchor_attribute_name: config.anchor_attribute_name,
          default_value: null
        },
        value
      )
      config.selectors[key].anchor_selector = normalizeSelector(
        config.selectors[key].anchor_selector
      )
    }
  })

  return config
}
