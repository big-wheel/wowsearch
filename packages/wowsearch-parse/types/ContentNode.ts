/**
 * @file ContentNode
 * @author Cuttle Cong
 * @date 2019/5/29
 *
 */

export default class ContentNode {
  public type: string
  public value: string
  public domNode?: Node

  constructor(value = null) {
    this.value = value
  }


}
