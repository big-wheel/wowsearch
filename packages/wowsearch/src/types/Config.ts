/**
 * @file Config
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */
import * as w from 'walli'
import * as each from 'lodash.foreach'
import * as clone from 'lodash.clonedeep'
import { Rule, walliRule } from '../match'

export type Selector =
  | string
  | {
      selector: string
      type?: 'xpath' | 'css'
      default_value?: any
      strip_chars?: string
    }

const walliSelector = w.oneOf([
  w.string,
  w.leq({
    selector: w.string,
    type: w.oneOf(['xpath', 'css']).optional,
    default_value: w.any.optional,
    strip_chars: w.string.optional
  })
])



export type CrawlerConfig = {
  js_render?: boolean
  js_waitfor?: string | number | Function
  strip_chars?: string
  start_urls?: Array<Rule | { url: Rule }>
  stop_urls?: Rule[]

  selectors: {
    lvl0?: Selector
    lvl1?: Selector
    lvl2?: Selector
    lvl3?: Selector
    lvl4?: Selector
    lvl5?: Selector
    lvl6?: Selector
    text?: Selector
  }

  selectors_exclude?: string[]
  smart_crawling?: boolean
  force_crawling_urls?: boolean
}

export type Config = CrawlerConfig & {
  sitemap_urls?: string[]
  sitemap_urls_patterns?: Rule[]
  force_sitemap_urls_crawling?: boolean
}

const WalliDef = w.leq({
  js_render: w.boolean.optional,
  js_waitfor: w.oneOf([w.string, w.number, w.function_]).optional,
  strip_chars: w.string.optional,
  start_urls: w.arrayOf(w.oneOf([walliRule, w.leq({ url: w.string })])).optional,
  stop_urls: w.arrayOf(walliRule).optional,

  selectors: {
    lvl0: walliSelector.optional,
    lvl1: walliSelector.optional,
    lvl2: walliSelector.optional,
    lvl3: walliSelector.optional,
    lvl4: walliSelector.optional,
    lvl5: walliSelector.optional,
    lvl6: walliSelector.optional,
    text: walliSelector.optional
  },

  selectors_exclude: w.arrayOf(w.string).optional
})

export function normalize(config) {
  let errorMsg = WalliDef.toUnlawfulString(config)
  if (errorMsg) {
    throw new TypeError(errorMsg)
  }
  config = clone(config)

  config = Object.assign(
    {
      strip_chars: ' .,;:§¶',
      js_render: false,
      js_waitfor: 0,
      start_urls: [/.*/],
      stop_urls: [],
      selectors_exclude: [],
      sitemap_urls: [],
      sitemap_urls_patterns: [/.*/],
      force_sitemap_urls_crawling: false,
      smart_crawling: true,
      force_crawling_urls: false
    },
    config
  )

  each(config.selectors, (value, key) => {
    if (typeof value === 'string') {
      value = {
        selector: value
      }
    }
    config.selectors[key] = Object.assign({ type: 'css', strip_chars: config.strip_chars, default_value: null }, value)
  })

  return config
}
