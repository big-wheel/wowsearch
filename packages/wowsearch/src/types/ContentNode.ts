/**
 * @file ContentNode
 * @author Cuttle Cong
 * @date 2019/5/29
 *
 */

export default class ContentNode {
  public type: string
  public value: string

  constructor(value = null) {
    this.value = value
  }


}
