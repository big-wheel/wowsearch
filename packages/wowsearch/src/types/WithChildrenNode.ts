import ContentNode from "./ContentNode";

/**
 * @file ContentNode
 * @author Cuttle Cong
 * @date 2019/5/29
 *
 */

export default class WithChildrenNode extends ContentNode {
  public children: ContentNode[]

  constructor(value = null, children = []) {
    super(value)
    this.children = children
  }
}
