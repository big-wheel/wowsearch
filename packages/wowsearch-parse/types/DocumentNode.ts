/**
 * @file ContentNode
 * @author Cuttle Cong
 * @date 2019/5/29
 *
 */

import WithChildrenNode from './WithChildrenNode'
import { Rule } from '../match'

export default class DocumentNode extends WithChildrenNode {
  public urlRule: Rule
  public global = new Map<string, string | string[]>()
  public readonly type = 'document'

  public href: string
}
