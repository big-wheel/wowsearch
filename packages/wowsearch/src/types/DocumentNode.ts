/**
 * @file ContentNode
 * @author Cuttle Cong
 * @date 2019/5/29
 *
 */

import WithChildrenNode from "./WithChildrenNode";

export default class DocumentNode extends WithChildrenNode {
  public global = new Map<string, string>()
  public readonly type = 'document'
}
