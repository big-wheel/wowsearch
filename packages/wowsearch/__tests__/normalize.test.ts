/**
 * @file main.test
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */
import { readFileSync } from 'fs'
import { makeFixture } from './help'
import { normalize } from '../src/types/Config'

describe('normalize', function() {
  it('normalize throw error', function() {
    expect(() => normalize({})).toThrowErrorMatchingSnapshot()
    expect(() =>
      normalize({ strip_chars: null, start_urls: 'abc' })
    ).toThrowErrorMatchingSnapshot()
  })
  it('should normalized', function() {
    let c = { selectors: {} }
    let normalized = normalize(c)
    expect(normalized).toMatchSnapshot()
    expect(normalized).not.toBe(c)
  })
})
