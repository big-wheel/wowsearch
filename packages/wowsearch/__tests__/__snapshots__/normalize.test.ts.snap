// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalize normalize throw error 1`] = `"selectors: expected type: object, actual type: undefined."`;

exports[`normalize normalize throw error 2`] = `
"strip_chars: expected type: string, actual type: null.
start_urls: expected type: function, actual type: string.
selectors: expected type: object, actual type: undefined."
`;

exports[`normalize should normalized 1`] = `
Object {
  "force_crawling_urls": false,
  "force_sitemap_urls_crawling": false,
  "js_render": false,
  "js_waitfor": 0,
  "selectors": Object {},
  "selectors_exclude": Array [],
  "sitemap_urls": Array [],
  "sitemap_urls_patterns": Array [],
  "smart_crawling": false,
  "start_urls": Array [
    "*",
  ],
  "stop_urls": Array [],
  "strip_chars": " .,;:§¶",
}
`;
